operators/pkg/controller/kibanaassociation/user.go

```diff
@@ -99,66 +96,63 @@ func reconcileEsUser(c k8s.Client, s *runtime.Scheme, kibana kbtype.Kibana, es t
 		Client:     c,
 		Scheme:     s,
 		Owner:      &kibana,
-		Expected:   &expectedSecret,
+		Expected:   &expectedKibanaSecret,
 		Reconciled: &reconciledSecret,
 		NeedsUpdate: func() bool {
-			_, ok := reconciledSecret.Data[kibanaUser]
-			return !ok || !hasExpectedLabels(&expectedSecret, &reconciledSecret)
+			_, ok := reconciledSecret.Data[username]
+			return !ok || !hasExpectedLabels(&expectedKibanaSecret, &reconciledSecret)
 		},
 		UpdateReconciled: func() {
-			setExpectedLabels(&expectedSecret, &reconciledSecret)
-			reconciledSecret.Data = expectedSecret.Data
+			setExpectedLabels(&expectedKibanaSecret, &reconciledSecret)
+			reconciledSecret.Data = expectedKibanaSecret.Data
 		},
 	})
 	if err != nil {
 		return err
 	}
-	reconciledPw := reconciledSecret.Data[kibanaUser] // make sure we don't constantly update the password
 
+	reconciledPw := reconciledSecret.Data[username] // make sure we don't constantly update the password
 	bcryptHash, err := bcrypt.GenerateFromPassword(reconciledPw, bcrypt.DefaultCost)
 	if err != nil {
 		return err
 	}
 
-	// analogous to the secret: the user goes on the Elasticsearch side of the association
+	// analogous to the Kibana secret: a user Secret goes on the Elasticsearch side of the association
 	// we apply the ES cluster labels ("user belongs to that ES cluster")
 	// and the association label ("for that Kibana association")
-	userLabels := label.NewLabels(es)
+	userLabels := commonuser.NewLabels(es)
 	userLabels[AssociationLabelName] = kibana.Name
 	usrKey := KibanaUserKey(kibana, es.Namespace)
-	expectedUser := &estype.User{
+
+	expectedEsUser := &corev1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      usrKey.Name,
 			Namespace: usrKey.Namespace,
 			Labels:    userLabels,
 		},
-		Spec: estype.UserSpec{
-			Name:         kibanaUser,
-			PasswordHash: string(bcryptHash),
-			UserRoles:    []string{user.KibanaSystemUserBuiltinRole},
-		},
-		Status: estype.UserStatus{
-			Phase: estype.UserPending,
+		Data: map[string][]byte{
+			commonuser.UserName:     []byte(username),
+			commonuser.PasswordHash: bcryptHash,
+			commonuser.UserRoles:    []byte(user.KibanaSystemUserBuiltinRole),
 		},
 	}
 
-	reconciledUser := estype.User{}
+	reconciledEsSecret := corev1.Secret{}
 	return reconciler.ReconcileResource(reconciler.Params{
 		Client:     c,
 		Scheme:     s,
 		Owner:      &kibana, // user is owned by the Kibana resource
```

Comment:
Not something you changed but I am wondering if this should be owned by the Elasticsearch resource as it might not be in the same namespace as the Kibana resource. 

I would make the additional watch logic superfluous. Downside would be that users created by APM/Kibana would not disappear automatically when the corresponding resource is deleted but that could be handled in the association controller I guess.

Reply:
I found an other bug in the controller while working on this one 😞 

Regarding your proposal I'm not sure it will be simpler to manage. If we do that we have to manage garbage collection manually for each association (APM and Kibana so far, may be more later ?) while it is done automatically for us by the gc even if the objects are not in the same namespace.

Reply:
User gc is now handled with a finalizer.